src/reference/add-wasm-support-to-crate.md

```diff
@@ -0,0 +1,156 @@
+
+## Adding Support for WebAssembly
+
+### Avoid Performing I/O Directly
+
+On the Web, I/O is always asynchronous, and there isn't a file system. Factor
+I/O out of your library, let users perform the I/O and then pass the input
+slices to your library instead.
+
+For example, refactor this:
+
+```rust
+use std::fs;
+use std::path::Path;
+
+pub fn parse_thing(path: &Path) -> Result<MyThing, MyError> {
+    let contents = fs::read(path)?;
+    // ...
+}
+```
+
+Into this:
+
+```rust
+pub fn parse_thing(contents: &[u8]) -> Result<MyThing, MyError> {
+    // ...
+}
+```
+
+### Add `wasm-bindgen` as a Dependency
+
+If you need to interact with the outside world (i.e. you can't have library
+consumers drive that interaction for you) then you'll need to add `wasm-bindgen`
+(and `js-sys` and `web-sys` if you need them) as a dependency for when
+compilation is targeting WebAssembly:
+
+```toml
+[target.'cfg(target_arch = "wasm32")'.dependencies]
+wasm-bindgen = "0.2"
+js-sys = "0.3"
+web-sys = "0.3"
+```
+
+### Avoid Synchronous I/O
+
+If you must perform I/O in your library, then it cannot be synchronous. There is
+only asynchronous I/O on the Web. Use [the `futures`
+crate](https://crates.io/crates/futures) and [the `wasm-bindgen-futures`
+crate](https://rustwasm.github.io/wasm-bindgen/api/wasm_bindgen_futures/) to
+manage asynchronous I/O. If you're library functions are generic over some
+future type `F`, then that future can be implemented via `fetch` on the Web or
+via non-blocking I/O provided by the operating system.
+
+```rust
+pub fn do_stuff<F>(future: F) -> impl Future<Item = MyOtherThing>
+where
+    F: Future<Item = MyThing>,
+{
+    // ...
+}
+```
+
+You can also define a trait and implement it for WebAssembly and the Web and
+also for native targets:
+
+```rust
+trait ReadMyThing {
+    type F: Future<Item = MyThing>;
+    fn read(&self) -> Self::F;
+}
+
+#[cfg(target_arch = "wasm32")]
+struct WebReadMyThing {
+    // ...
+}
+
+#[cfg(target_arch = "wasm32")]
+impl ReadMyThing for WebReadMyThing {
+    // ...
+}
+
+#[cfg(not(target_arch = "wasm32"))]
+struct NativeReadMyThing {
+    // ...
+}
+
+#[cfg(not(target_arch = "wasm32"))]
+impl ReadMyThing for NativeReadMyThing {
+    // ...
+}
+```
+
+### Avoid Spawning Threads
+
+You can use `#[cfg(..)]`s to enable threaded and non-threaded code paths
+depending on if the target is WebAssembly or not:
+
+```rust
+#![cfg(target_arch = "wasm32")]
+fn do_work() {
+    // Do work with only this thread...
+}
+
+#![cfg(not(target_arch = "wasm32"))]
+fn do_work() {
+    use std::thread;
+
+    // Spread work to helper threads....
+    thread::spawn(|| {
+        // ...
+    });
+}
+```
+
+## Maintaining Ongoing Support for WebAssembly
+
+### Building for `wasm32-unknown-unknown` in CI
+
+Ensure that compilation doesn't fail when targeting WebAssembly by having your
+CI script run these commands:
+
+```
+rustup target add wasm32-unknown-unknown
+cargo check --target wasm32-unknown-unknown
+```
```

Comment:
It may be a good idea to mention how to do this in Travis:

```yml
matrix:
  include:
    - rust: stable
      name: "check wasm32 support"
      install: rustup target add wasm32-unknown-unknown
      script: cargo check --target wasm32-unknown-unknown